### PR TITLE
Workflow audit code

### PR DESCRIPTION
**Pull request by johntran777**
Please review and merge.

### CODE REVIEW

**Comment by minitru** [comment]
+1

File: pp-api-gateway-model/src/main/java/com/openenglish/pp/persistence/entity/Workflow.java [hunk 1]
```diff
@@ -56,6 +59,7 @@ public WorkflowStatus getStatus() {
   }
 
   public Workflow setStatus(WorkflowStatus status) {
+    Preconditions.checkNotNull(status, "status can not null!");
```
Comment (ecaravatti) [inline pp-api-gateway-model/src/main/java/com/openenglish/pp/persistence/entity/Workflow.java]: You missed the verb :P

File: pp-api-gateway-model/src/main/liquibase/changesets/baselines/sql/db.create.sql [hunk 2]
```diff
@@ -23,4 +23,40 @@ CACHE 1;
 ALTER SEQUENCE WORKFLOW_ID_SEQ OWNED BY WORKFLOW.WORKFLOW_ID;
 
 ALTER TABLE ONLY WORKFLOW ALTER COLUMN WORKFLOW_ID SET DEFAULT nextval('WORKFLOW_ID_SEQ'::regclass);
-ALTER TABLE ONLY WORKFLOW ADD CONSTRAINT PK_WORKFLOW PRIMARY KEY (WORKFLOW_ID);
\ No newline at end of file
+ALTER TABLE ONLY WORKFLOW ADD CONSTRAINT PK_WORKFLOW PRIMARY KEY (WORKFLOW_ID);
```
Comment (ecaravatti) [inline pp-api-gateway-model/src/main/liquibase/changesets/baselines/sql/db.create.sql]: Normally, I'd say that you shouldn't modify an sql file since it will alter the checksum and liquibase would fail. But since it's already merged, not sure if you want to revert the changes to this file and create a new one, or clean the checksums in the db.

File: pp-api-gateway-model/src/main/liquibase/changesets/baselines/sql/db.create.sql [hunk 3]
```diff
@@ -23,4 +23,40 @@ CACHE 1;
...
+CACHE 1;
+
+ALTER SEQUENCE WORKFLOW_AUDIT_ID_SEQ OWNED BY WORKFLOW_AUDIT.WORKFLOW_AUDIT_ID;
+
+ALTER TABLE ONLY WORKFLOW_AUDIT ALTER COLUMN WORKFLOW_AUDIT_ID SET DEFAULT nextval('WORKFLOW_AUDIT_ID_SEQ'::regclass);
+ALTER TABLE ONLY WORKFLOW_AUDIT ADD CONSTRAINT PK_WORKFLOW_AUDIT PRIMARY KEY (WORKFLOW_AUDIT_ID);
+CREATE INDEX IDX_WORKFLOW_AUDIT_WF_ID ON WORKFLOW_AUDIT USING btree (WORKFLOW_ID);
+CREATE INDEX IDX_WORKFLOW_AUDIT_STEP ON WORKFLOW_AUDIT USING btree (STEP);
+CREATE INDEX IDX_WORKFLOW_AUDIT_CREATED_DATE ON WORKFLOW_AUDIT USING btree (CREATED_DATE);
+CREATE INDEX IDX_WORKFLOW_AUDIT_WF_ID_STATUS_CREATED ON WORKFLOW_AUDIT USING btree (WORKFLOW_ID, STATUS, CREATED_DATE);
```
Comment (ecaravatti) [inline pp-api-gateway-model/src/main/liquibase/changesets/baselines/sql/db.create.sql]: hmmm... I'm not so sure about having as many indexes here, this may impact inserts on the table, which should be frequent enough. We may want to keep an eye on its performance.
And we should probably consider partitioning this table as it will keep growing indefinitely.

File: pp-api-gateway-model/src/test/java/com/openenglish/pp/persistence/repository/WorkflowAuditRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+@DatabaseSetup(value = "workflowAuditData.xml", type = DatabaseOperation.INSERT)
+@DatabaseTearDown(value = "workflowAuditData.xml", type = DatabaseOperation.DELETE)
+public class WorkflowAuditRepositoryTest extends AbstractPersistenceTest {
+
+  @Autowired
+  private WorkflowAuditRepository workflowAuditRepository;
+
+
+  @Before
```
Comment (ecaravatti) [inline pp-api-gateway-model/src/test/java/com/openenglish/pp/persistence/repository/WorkflowAuditRepositoryTest.java]: If we're not doing any setup or teardown steps, we could delete these methods.

File: pp-api-gateway-service/src/main/java/com/openenglish/pp/service/flow/ImmediatePaymentFlow.java [hunk 5]
```diff
@@ -167,16 +170,16 @@ private boolean handleAccountSubscribePayZoura(ImmediatePaymentMessage message)
                                                 message.getLeadId(), message.getProductRatePlanChargeId(),
                                                 message.getCurrency(), message.getCreditCardToken());
         logger.error(failedPaymentMsg, ex);
```
Comment (ecaravatti) [inline pp-api-gateway-service/src/main/java/com/openenglish/pp/service/flow/ImmediatePaymentFlow.java]: I believe you could format the message directly in the logger doing
```java
logger.error("Failed capturing payment for leadId: {}, productRatePlanId:{}, currency:{}, 
        cardtoken:{}", message.getLeadId(), message.getProductRatePlanChargeId(), 
        message.getCurrency(), message.getCreditCardToken(), ex);
```

File: pp-api-gateway-service/src/main/java/com/openenglish/pp/service/util/WorkflowUtil.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+@Component
+public class WorkflowUtil {
+
+  @Autowired
+  private WorkflowRepository workflowRepository;
+  @Autowired
+  private WorkflowAuditRepository workflowAuditRepository;
+
+  public Workflow recordWorkflowStart(String step, Object auditMessage) {
```
Comment (ecaravatti) [inline pp-api-gateway-service/src/main/java/com/openenglish/pp/service/util/WorkflowUtil.java]: This should be @transactional.

File: pp-api-gateway-service/src/main/java/com/openenglish/pp/service/util/WorkflowUtil.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  private void createWorkflowAudit(Workflow workflow, String step, Object auditMessage) {
+    WorkflowAudit workflowAudit = new WorkflowAudit()
+                                      .setWorkflowId(workflow.getId())
+                                      .setStep(step)
+                                      .setMessage(StringUtils.substring(auditMessage.toString(), 0, 255));
+    workflowAuditRepository.save(workflowAudit);
+  }
+
+
+  public Workflow recordWorkflowProgress(Long workflowId, WorkflowStatus status, String step, Object auditMessage) {
```
Comment (ecaravatti) [inline pp-api-gateway-service/src/main/java/com/openenglish/pp/service/util/WorkflowUtil.java]: This should be @transactional.
